### PR TITLE
Add nvram, redirfilter and shmem device support

### PR DESCRIPTION
**Pull request by chunfuwen**
This is subsequent follow-ups requests from hanhan in PR: https://github.com/avocado-framework/avocado-vt/pull/3056

Signed-off-by: chunfuwen <chwen@redhat.com>

### CODE REVIEW

**Comment by chunfuwen** [comment]
>>> from virttest.libvirt_xml.devices.redirfilter import Redirfilter
>>> aa = Redirfilter()
>>> input_attrs1 = {'class':'0x08', 'vendor':'0x1234', 'product':'0xbeef', 'version':'2.56', 'allow':'yes'}
>>> input_attrs2 = {'class':'0x04', 'vendor':'0x1234', 'product':'0xbeef', 'version':'2.56', 'allow':'no'}
>>> usbdev_list = []
>>> usbdev_list.append(input_attrs1)
>>> usbdev_list.append(input_attrs2)
>>> setattr(aa, 'usbdevs', usbdev_list)
>>> print(aa)
<?xml version='1.0' encoding='UTF-8'?>
<redirfilter><usbdev allow="yes" class="0x08" product="0xbeef" vendor="0x1234" version="2.56" /><usbdev allow="no" class="0x04" product="0xbeef" vendor="0x1234" version="2.56" /></redirfilter>
>>> 


>> aa = Nvram()
>>> setattr(aa, 'address_type', 'spapr-vio')
>>> setattr(aa, 'address_reg', '0x00003000')
>>> print(aa)
<?xml version='1.0' encoding='UTF-8'?>
<nvram><address reg="0x00003000" type="spapr-vio" /></nvram>
>>> 


>>> from virttest.libvirt_xml.devices.shmem import Shmem
>>> aa = Shmem()
>>> aa.name = 'my_shmem0'
>>> aa.role = 'peer'
>>> model_attrs = {'type':'ivshmem-plain'}
>>> aa.model_attrs = model_attrs
>>> aa.size=2
>>> aa.size_unit = 'M'
>>> server_attrs = {'path':'/tmp/socket-shmem'}
>>> aa.server_attrs = server_attrs
>>> msi_attrs={'vectors':'32', 'ioeventfd':'on'}
>>> aa.msi_attrs = msi_attrs
>>> print(aa)
<?xml version='1.0' encoding='UTF-8'?>
<shmem name="my_shmem0" role="peer"><model type="ivshmem-plain" /><size unit="M">2</size><server path="/tmp/socket-shmem" /><output><msi ioeventfd="on" vectors="32" /></output></shmem>
>>>

**Comment by chunfuwen** [comment]
```
>>> from virttest.libvirt_xml.devices.redirfilter import Redirfilter
>>> aa = Redirfilter()
>>> input_attrs1 = {'class':'0x08', 'vendor':'0x1234', 'product':'0xbeef', 'version':'2.56', 'allow':'yes'}
>>> input_attrs2 = {'class':'0x04', 'vendor':'0x1234', 'product':'0xbeef', 'version':'2.56', 'allow':'no'}
>>> usbdev_list = []
>>> usbdev_list.append(input_attrs1)
>>> usbdev_list.append(input_attrs2)
>>> setattr(aa, 'usbdevs', usbdev_list)
>>> print(aa)
<?xml version='1.0' encoding='UTF-8'?>
<redirfilter><usbdev allow="yes" class="0x08" product="0xbeef" vendor="0x1234" version="2.56" /><usbdev allow="no" class="0x04" product="0xbeef" vendor="0x1234" version="2.56" /></redirfilter>
>>> 


>> aa = Nvram()
>>> setattr(aa, 'address_type', 'spapr-vio')
>>> setattr(aa, 'address_reg', '0x00003000')
>>> print(aa)
<?xml version='1.0' encoding='UTF-8'?>
<nvram><address reg="0x00003000" type="spapr-vio" /></nvram>
>>> 


>>> from virttest.libvirt_xml.devices.shmem import Shmem
>>> aa = Shmem()
>>> aa.name = 'my_shmem0'
>>> aa.role = 'peer'
>>> model_attrs = {'type':'ivshmem-plain'}
>>> aa.model_attrs = model_attrs
>>> aa.size=2
>>> aa.size_unit = 'M'
>>> server_attrs = {'path':'/tmp/socket-shmem'}
>>> aa.server_attrs = server_attrs
>>> msi_attrs={'vectors':'32', 'ioeventfd':'on'}
>>> aa.msi_attrs = msi_attrs
>>> print(aa)
<?xml version='1.0' encoding='UTF-8'?>
<shmem name="my_shmem0" role="peer"><model type="ivshmem-plain" /><size unit="M">2</size><server path="/tmp/socket-shmem" /><output><msi ioeventfd="on" vectors="32" /></output></shmem>
>>> 



```

**Comment by chunfuwen** [comment]
@kylazhang 
@hahakiki2010 
please help take a look

**Review by kylazhang — APPROVED** [review]

**Review by kylazhang — APPROVED** [review]

File: virttest/libvirt_xml/devices/shmem.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+"""
+shared memory device support class(es)
+
+<devices>
+  <shmem name='my_shmem0' role='peer'>
+    <model type='ivshmem-plain'/>
```
Comment (qiankehan) [inline virttest/libvirt_xml/devices/shmem.py:6]: `<address type="pci" ...>` is optional for shmem device, too. See the example of XML: https://gitlab.com/libvirt/libvirt/-/blob/master/tests/qemuxml2argvdata/shmem-plain-doorbell.xml#L32
@chunfuwen Please add that in the code comment and `__init__()`

**Review by qiankehan — CHANGES_REQUESTED** [review]
See the previous comment. Others look good to me.

File: virttest/libvirt_xml/devices/shmem.py (hunk 1 above)
Comment (chunfuwen) [reply virttest/libvirt_xml/devices/shmem.py:6]: @qiankehan 
updated as per comments

Comment (chunfuwen) [reply virttest/libvirt_xml/devices/shmem.py:6]: @qiankehan
updated as per comments

Comment (chunfuwen) [reply virttest/libvirt_xml/devices/shmem.py:6]: @qiankehan
updated as per comments

**Comment by chunfuwen** [comment]
validate latest change result:
```
>>> from virttest.libvirt_xml.devices.shmem import Shmem
>>> aa = Shmem()
>>> aa.name = 'my_shmem0'
>>> model_attrs = {'type':'ivshmem-plain'}
>>> aa.model_attrs = model_attrs
>>> pci_address_attrs = {'type':'pci', 'domain':'0x0000', 'bus':'0x00', 'slot':'0x04', 'function':'0x0'}
>>> aa.pci_addr_attrs =pci_address_attrs
>>> aa
{'xmltreefile': None, 'validates': None, 'xml': <virttest.xml_utils.XMLTreeFile object at 0x7fb827d39c50>, 'virsh': <module 'virttest.virsh' from '/root/workdir/avocado-vt/virttest/virsh.py'>, 'device_tag': 'shmem'}
>>> print(aa)
<?xml version='1.0' encoding='UTF-8'?>
<shmem name="my_shmem0"><model type="ivshmem-plain" /><address bus="0x00" domain="0x0000" function="0x0" slot="0x04" type="pci" /></shmem>
>>> 

```

File: virttest/libvirt_xml/devices/shmem.py (hunk 1 above)
Comment (qiankehan) [reply virttest/libvirt_xml/devices/shmem.py:6]: LGTM

**Review by qiankehan — APPROVED** [review]

**Pull request merged**